src/back/spv/writer.rs

```diff
@@ -2098,9 +2109,147 @@ impl Writer {
                 });
                 id
             }
-            crate::Expression::ImageQuery { .. }
-            | crate::Expression::Relational { .. }
-            | crate::Expression::ArrayLength(_) => {
+            crate::Expression::ImageQuery { image, query } => {
+                use crate::{ImageClass as Ic, ImageDimension as Id, ImageQuery as Iq};
+
+                let image_id = self.get_expression_global(ir_function, image);
+                let image_type = fun_info[image].ty.handle().unwrap();
+                let (dim, arrayed, class) = match ir_module.types[image_type].inner {
+                    crate::TypeInner::Image {
+                        dim,
+                        arrayed,
+                        class,
+                    } => (dim, arrayed, class),
+                    _ => {
+                        return Err(Error::Validation("image type"));
+                    }
+                };
+
+                match query {
+                    Iq::Size { level } => {
+                        let dim_coords = match dim {
+                            Id::D1 => 1,
+                            Id::D2 | Id::Cube => 2,
+                            Id::D3 => 3,
+                        };
+                        let extended_size_type_id = {
+                            let array_coords = if arrayed { 1 } else { 0 };
+                            let vector_size = match dim_coords + array_coords {
+                                2 => Some(crate::VectorSize::Bi),
+                                3 => Some(crate::VectorSize::Tri),
+                                4 => Some(crate::VectorSize::Quad),
+                                _ => None,
```

Comment:
This should be `unreachable!`, it seems to me, and the `Some` should be in the `LocalType::Value` constructor.

Reply:
It's valid to be `None` here, which happens if you query a 1D image size (non-arrayed)